test/lit/passes/remove-unused-brs_all-features.wast

```diff
@@ -142,33 +153,115 @@
    (drop
     (br_on_null $null (ref.func $br_on_null))
    )
+   ;; If we don't know whether the input is null, we can't optimize.
+   (drop
+    (br_on_null $null (local.get $x))
+   )
   )
  )
 
- ;; CHECK:      (func $br_on_non_null (type $none_=>_funcref) (result funcref)
- ;; CHECK-NEXT:  (block $non-null (result (ref $none_=>_funcref))
+ ;; CHECK:      (func $br_on_null-fallthrough (type $none_=>_none)
+ ;; CHECK-NEXT:  (local $x funcref)
+ ;; CHECK-NEXT:  (block $null
+ ;; CHECK-NEXT:   (drop
+ ;; CHECK-NEXT:    (block
+ ;; CHECK-NEXT:     (drop
+ ;; CHECK-NEXT:      (local.tee $x
+ ;; CHECK-NEXT:       (ref.null nofunc)
+ ;; CHECK-NEXT:      )
+ ;; CHECK-NEXT:     )
+ ;; CHECK-NEXT:     (br $null)
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (drop
+ ;; CHECK-NEXT:    (ref.as_non_null
+ ;; CHECK-NEXT:     (local.tee $x
+ ;; CHECK-NEXT:      (ref.func $br_on_null)
+ ;; CHECK-NEXT:     )
+ ;; CHECK-NEXT:    )
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:  )
+ ;; CHECK-NEXT: )
+ (func $br_on_null-fallthrough
+  ;; This is the same as above, but now the necessary type information comes
+  ;; from fallthrough values.
+  (local $x funcref)
+  (block $null
+   ;; Definitely taken.
+   (drop
+    (br_on_null $null (local.tee $x (ref.null nofunc)))
+   )
+   ;; Definitely not taken. Optimizable, but still requires a cast for validity.
+   (drop
+    (br_on_null $null (local.tee $x (ref.func $br_on_null)))
+   )
+  )
+ )
+
+ ;; CHECK:      (func $br_on_non_null (type $funcref_=>_funcref) (param $x funcref) (result funcref)
+ ;; CHECK-NEXT:  (block $non-null (result (ref func))
  ;; CHECK-NEXT:   (br $non-null
  ;; CHECK-NEXT:    (ref.func $br_on_non_null)
  ;; CHECK-NEXT:   )
- ;; CHECK-NEXT:   (br_on_non_null $non-null
+ ;; CHECK-NEXT:   (drop
  ;; CHECK-NEXT:    (ref.null nofunc)
  ;; CHECK-NEXT:   )
- ;; CHECK-NEXT:   (ref.func $br_on_non_null)
+ ;; CHECK-NEXT:   (br_on_non_null $non-null
+ ;; CHECK-NEXT:    (local.get $x)
+ ;; CHECK-NEXT:   )
+ ;; CHECK-NEXT:   (unreachable)
  ;; CHECK-NEXT:  )
  ;; CHECK-NEXT: )
- (func $br_on_non_null (result funcref)
+ (func $br_on_non_null (param $x funcref) (result funcref)
   (block $non-null (result (ref func))
    ;; A non-null reference is not null, and the br is always taken.
    (br_on_non_null $non-null
     (ref.func $br_on_non_null)
    )
    ;; On the other hand, if we know the input is null, the branch will never be
    ;; taken.
-   ;; TODO: Optimize this.
    (br_on_non_null $non-null
     (ref.null nofunc)
    )
-   (ref.func $br_on_non_null)
+   ;; If we don't know whether the input is null, we can't optimize.
+   (br_on_non_null $non-null
+    (local.get $x)
+   )
+   (unreachable)
+  )
+ )
+
+ ;; CHECK:      (func $br_on_non_null-fallthrough (type $none_=>_funcref) (result funcref)
+ ;; CHECK-NEXT:  (local $x funcref)
+ ;; CHECK-NEXT:  (block $non-null (result (ref func))
+ ;; CHECK-NEXT:   (br $non-null
+ ;; CHECK-NEXT:    (ref.as_non_null
```

Comment:
Similar question as before, we turned `br_on_non_null` into `br + ref.as_non_null` which seems larger and not faster?

Reply:
Similarly, we know for a fact that the `ref.as_non_null` will succeed at runtime, so we are expecting other optimizations to be able to get rid of it. Even if they don't, making the branch unconditional will probably allow  following code to be DCE'd, so I would still expect this to be beneficial.